Sources/OktaAnalytics/OktaAnalytics.swift

```diff
@@ -111,44 +143,138 @@ public final class OktaAnalytics: NSObject {
         }
+    }
 
-        func readLock() {
-            pthread_rwlock_rdlock(&self.lock)
+    /**
+        update a property scenario with the values.
+
+        - Parameters:
+           - scenarioID: unique sceario ID returned from `startScenario(_ , _)` .
+           - propertySubject: A closure that takes a `PassthroughSubject` of `Property` objects as a parameter.
+        */
+    public static func updateScenario(_ scenarioID: ScenarioID, _ properties: [Property]) {
+        storage.insertScenarioProperties(properties.compactMap { ScenarioProperty(scenarioID: scenarioID, key: $0.key, value: $0.value) })
+    }
+
+    public static func getScenarioEventByID(_ scenarioID: ScenarioID, _ completion: @escaping (ScenarioEvent?) -> Void) {
+        storage.fetchScenarioAndProperties(scenarioID) { scenario, scenarioProperties in
+            guard let scenario = scenario else {
+                completion(nil)
+                Self.providers.forEach {
+                    $1.logger?.log(level: .debug, eventName: scenarioID, message: "Failed to fetch scenario event by \(scenarioID)", properties: nil, file: #file, line: #line, funcName: #function)
+                }
+                return
+            }
+            completion(ScenarioEvent(name: scenario.name, displayName: scenario.displayName, properties: scenarioProperties.compactMap { Property(key: $0.key, value: $0.value) }))
         }
+    }
 
-        func unlock() {
-            pthread_rwlock_unlock(&self.lock)
+    public static func getOngoingScenarioIds(_ scenarioName: Name, _ completion: @escaping ([ScenarioID]) -> Void) {
+        storage.fetchScenarios(by: scenarioName) {
+            completion($0.compactMap { $0.id })
         }
+    }
+
+    /**
+        end sceanrio with ID. send properties to provider and clears local storage.
 
-        deinit {
-            pthread_rwlock_destroy(&self.lock)
+        - Parameters:
+           - scenarioID: unique sceario ID returned from `startScenario(_ , _)` .
+           - eventDisplayName: event name to display on provider dashboard.
+        */
+    public static func endScenario(_ scenarioID: ScenarioID, eventDisplayName: Name) {
+        storage.fetchScenarioAndProperties(scenarioID) { scenario, properties in
+            guard let scenario = scenario else {
+                Self.providers.forEach {
+                    $1.logger?.log(level: .debug, eventName: scenarioID, message: "Failed to end scenario \(scenarioID) because scenario won't exists", properties: nil, file: #file, line: #line, funcName: #function)
+                }
+                return
+            }
+            var propertiesDict: Dictionary = .init(uniqueKeysWithValues: properties.lazy.map { ($0.key, $0.value) })
+            propertiesDict["DurationMS"] = "\(scenario.startTime.distance(to: Date()) * 1000)"
+            trackEvent(eventDisplayName, withProperties: propertiesDict)
+            storage.deleteScenariosByIds([scenarioID])
```

Comment:
99% sure that this will cause deadlock in GRDB since you are trying to write from read completion

Reply:
separated reads and writes